apps/query-eval/queryeval/driver.py

```diff
@@ -371,11 +400,38 @@ def do_eval(self, query: QueryEvalQuery, result: QueryEvalResult) -> QueryEvalRe
                 console.print("[red]:x: Documents retrieved don't include all expected docs")
                 console.print(f"Missing docs: {expected_doc_set - retrieved_doc_set})")
             metrics.doc_retrieval_recall = len(retrieved_doc_set & expected_doc_set) / len(expected_doc_set)
-            metrics.doc_retrieval_precision = len(retrieved_doc_set & expected_doc_set) / len(result.retrieved_docs)
+            metrics.doc_retrieval_precision = len(retrieved_doc_set & expected_doc_set) / len(retrieved_doc_set)
 
-        # Evaluate result
-        if not result.result:
-            console.print("[yellow] No query execution result available, skipping..", style="italic")
+        # Evaluate string metrics
+        if not query.expected:
```

Comment:
this function is large enough now that it might make sense to separate the specific metric types into separate methods (doc_retrieval metrics, generated_answer metrics)

Reply:
I have made these changes.